File: 383.ransom-note.java
/*
 * @lc app=leetcode id=383 lang=java
 *
 * [383] Ransom Note
 */

// @lc code=start

import java.util.Arrays;

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] records = new int[26];
        for (char ch : magazine.toCharArray()) {
            records[ch - 'a'] += 1;
        }
        for (char ch : ransomNote.toCharArray()) {
            records[ch - 'a'] -= 1;
        }
        return Arrays.stream(records)
                .filter(x -> x < 0)
                .count() == 0;
    }
}
// @lc code=end

File: 383.ransom-note.cpp
/*
 * @lc app=leetcode id=383 lang=cpp
 *
 * [383] Ransom Note
 */

// @lc code=start

#include <string>
#include <array>

using namespace std;

class Solution {
public:
    using table_type = array<int, 26>;
    bool canConstruct(string ransomNote, string magazine) {
        table_type rt = buildTable(ransomNote), mt = buildTable(magazine);
        transform(rt.begin(), rt.end(), mt.begin(), mt.begin(), [](const auto a, const auto b) {
            return b - a;
        });
        return all_of(mt.begin(), mt.end(), [](const auto e) {
            return e >= 0;
        });
    }
    table_type buildTable(const string &s) {
        table_type t;
        for (auto ch: s) {
            t[ch - 'a'] += 1;
        }
        return t;
    }
};
// @lc code=end
